### ISSUE TITLE
Failed on building wrapper functions at CUDA 9.0

### ISSUE THREAD

**Issue by iSergio** [heterodb/pg-strom#307]
2017-10-14 15:37:11.982 MSK [6532] LOG:  PG-Strom version 2.0devel built for PostgreSQL 9.6
2017-10-14 15:37:11.997 MSK [6532] LOG:  PG-Strom: GPU0 GeForce GTX 1060 6GB (1280 CUDA cores; 1759MHz, L2 1536kB), RAM 5.93GB (192bits, 3.82GHz), CC 6.1
2017-10-14 15:37:11.998 MSK [6532] LOG:  NVRTC - CUDA Runtime Compilation vertion 9.0
gcc: error trying to exec 'cc1plus': execvp: No such file or directory
2017-10-14 15:37:12.003 MSK [6532] FATAL:  failed on nvcc (/usr/local/cuda/bin/nvcc  --relocatable-device-code=true --gpu-architecture=sm_61 -DPGSTROM_BUILD_WRAPPER -I /usr/share/postgresql/9.6/extension --device-c /tmp/ko9wrD.cu -o /tmp/ko9wrD.cu.sm_61.o)
2017-10-14 15:37:12.003 MSK [6532] LOG:  database system is shut down

All gcc c++ and etc installed
Command /usr/local/cuda/bin/nvcc  --relocatable-device-code=true --gpu-architecture=sm_61 -DPGSTROM_BUILD_WRAPPER -I /usr/share/postgresql/9.6/extension --device-c /tmp/ko9wrD.cu -o /tmp/ko9wrD.cu.sm_61.o worked separated from root and postgres users

**Comment by kaigai** [comment]
Hmm... CUDA9.0 might change something.

Because of some GPU kernel functions (like ```__ballot()```) got deprecated at CUDA 9.0,
I didn't test the code based on CUDA 9.0.
Let me check.

**Comment by kaigai** [comment]
Sorry, I couldn't reproduce the problem on CUDA9.0.

Could you confirm the ```cc1plus``` package is successfully installed.
(run ```rpm -ql gcc-c++```)

Also, is it possible to confirm the ```PATH``` environment variable of the postmaster process?

**Comment by iSergio** [comment]
Ok, i'm tested. This happens on Ubuntu 16.XX Server. Start PostgreSQL cluster through pg_ctlcluster writes on perl for debian systems. If start cluster through:
`
pg_ctl -D /media/database/postgresql/9.6/main -o "-c config_file=/etc/postgresql/9.6/main/postgresql.conf" start
`
I get:
`
2017-10-20 09:37:16.263 MSK [26075] LOG:  PG-Strom: GPU0 GeForce GTX 1060 6GB (1280 CUDA cores; 1759MHz, L2 1536kB), RAM 5.93GB (192bits, 3.82GHz), CC 6.1
2017-10-20 09:37:16.263 MSK [26075] LOG:  NVRTC - CUDA Runtime Compilation vertion 9.0
2017-10-20 09:37:20.974 MSK [26153] LOG:  database system was shut down at 2017-10-20 09:29:34 MSK
2017-10-20 09:37:20.977 MSK [26153] LOG:  MultiXact member wraparound protections are now enabled
2017-10-20 09:37:20.981 MSK [26157] LOG:  autovacuum launcher started
2017-10-20 09:37:20.981 MSK [26075] LOG:  database system is ready to accept connections
`
How can you see, cluster was start, but on create extension i get error:
`
geobase_v7=# CREATE EXTENSION pg_strom;
ERROR:  could not find function "pgstrom_dma_buffer_info" in file "/usr/lib/postgresql/9.6/lib/pg_strom.so"
`
I will continue the experiments

**Comment by kaigai** [comment]
Installation SQL files were not updated to follow the latest implementation.
I removed deprecated definitions, so the latest ```master``` will work.

**Comment by dimon777** [comment]
I see the same issue on Postgres 10, Ubuntu 16.04 and Cuda 9.0. I suspect this is due to the way postgres service starts during: `service postgresql start`, namely it un-taint the environment making execvp not available during runtime compillation. However I tried using (in  /lib/systemd/system/postgresql@.service):
```
[Service]
Environment="PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/cuda/bin"
```
and this didn't help.

I also tried to comment relevant lines in /usr/bin/pg_ctlcluster
```
# untaint environment
$ENV{'PATH'} = '/sbin:/bin:/usr/sbin:/usr/bin';
delete @ENV{'IFS', 'CDPATH', 'ENV', 'BASH_ENV'};
```
which also didn't work. I am out of ideas at a moment how to fix this.

**Comment by kaigai** [comment]
Please use CUDA9.1.
http://heterodb.github.io/pg-strom/install/#checklist

CUDA9.0 or older version has problems to build code which includes half_t type.

**Comment by iSergio** [comment]
Issue not in build pg-strom. Issue in start cluster through pg_ctlcluster, not pass compilation:
`gcc: error trying to exec 'cc1plus': execvp: No such file or directory
2017-10-14 15:37:12.003 MSK [6532] FATAL: failed on nvcc (/usr/local/cuda/bin/nvcc --relocatable-device-code=true --gpu-architecture=sm_61 -DPGSTROM_BUILD_WRAPPER -I /usr/share/postgresql/9.6/extension --device-c /tmp/ko9wrD.cu -o /tmp/ko9wrD.cu.sm_61.o)`

**Comment by SlowBroshka** [comment]
Hi guys? i have this problem too? i am using clear ubuntu 16.04 and Cuda 9.1
`gcc: error trying to exec 'cc1plus': execvp: No such file or directory
2018-04-25 01:07:16.312 MSK [23903] FATAL:  failed on nvcc (/usr/local/cuda/bin/nvcc  --relocatable-device-code=true --gpu-architecture=sm_61 -DPGSTROM_BUILD_WRAPPER -I /usr/share/postgresql/10/extension --device-c /tmp/Pa0Mm3.cu -o /tmp/Pa0Mm3.cu.sm_61.o)`

**Comment by kaigai** [comment]
It may be a similar case:
https://stackoverflow.com/questions/16466488/cuda-cant-compile-cu-files-because-of-gcc-error-trying-to-exec-cc1plus

Does `apt-get install g++` help the problem?

**Comment by iSergio** [comment]
Problem not in this. After compile pgstrom and install them, cluster can be started through pg_ctl. But through pg_ctlcluster not

**Comment by kaigai** [comment]
PG-Strom tries to call `nvcc` on its startup to build own wrapper functions, not build time.
We cannot build the wrapper program on binary build time because it is not predictable which GPU model is used.

**Comment by SlowBroshka** [comment]
I create symbolic kink on cc1plus (`ln -s /usr/lib/gcc/x86_64-linux-gnu/5/cc1plus /usr/bin/cc1plus
`) but it not solve the problem. and then error was: `2018-04-28 21:43:34.723 MSK [21601] LOG:  NVRTC - CUDA Runtime Compilation vertion 9.1
sh: 1: gcc: not found
2018-04-28 21:43:34.729 MSK [21601] FATAL:  failed on nvcc (/usr/local/cuda/bin/nvcc  --relocatable-device-code=true --gpu-architecture=sm_61 -DPGSTROM_BUILD_WRAPPER -I /usr/share/postgresql/10/extension --device-c /tmp/eeatF9.cu -o /tmp/eeatF9.cu.sm_61.o)` but i have gcc

**Comment by SlowBroshka** [comment]
I use `cat /opt/PostgreSQL/10/data/postmaster.pid` to find PID (first line in output) and then do `cat /proc/<PID of postmaster>/environ`. 
My output: 
`LC_PAPER=ru_RU.UTF-8MANPATH=:/opt/PostgreSQL/10/share/manLC_ADDRESS=ru_RU.UTF-8LC_MONETARY=ru_RU.UTF-8TERM=xterm-256colorSHELL=/bin/bashLC_NUMERIC=ru_RU.UTF-8OLDPWD=/opt/PostgreSQL/10USER=postgresPGPORT=5432LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:LC_TELEPHONE=ru_RU.UTF-8SUDO_USER=userSUDO_UID=1000PGUSER=postgresPGDATABASE=postgresUSERNAME=rootMC_TMPDIR=/tmp/mc-postgresPATH=/opt/PostgreSQL/10/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/gamesMAIL=/var/mail/postgresLC_IDENTIFICATION=ru_RU.UTF-8PWD=/opt/PostgreSQL/10/binPGLOCALEDIR=/opt/PostgreSQL/10/share/localeLANG=en_US.UTF-8LC_MEASUREMENT=ru_RU.UTF-8HISTCONTROL=ignorespaceHOME=/opt/PostgreSQL/10SUDO_COMMAND=/bin/su postgresSHLVL=2MC_SID=15746LOGNAME=postgresPGDATA=/opt/PostgreSQL/10/dataSUDO_GID=1000LC_TIME=ru_RU.UTF-8LC_NAME=ru_RU.UTF-8_=./postgres`

**Comment by kaigai** [comment]
Hmm... It's strange.

Can you run the `nvcc` command which is displayed in the log message on your shell?
```
/usr/local/cuda/bin/nvcc \
  --relocatable-device-code=true \
  --gpu-architecture=sm_61 \
  -DPGSTROM_BUILD_WRAPPER \
  -I /usr/share/postgresql/10/extension \
  --device-c /tmp/eeatF9.cu -o /tmp/eeatF9.cu.sm_61.o
```
If you comment out `unlink(src_fname)` in line 1594 and 1601 of `src/cuda_program.c`, the source `*.cu` file in the temporary directory will be kept after the error.

And, can you find out any other message which might be generated by `nvcc` in the log files?

Thanks,

**Comment by SlowBroshka** [comment]
I exec command from log message:
`/usr/local/cuda/bin/nvcc  --relocatable-device-code=true --gpu-architecture=sm_61 -DPGSTROM_BUILD_WRAPPER -I /usr/share/postgresql/10/extension --device-c /tmp/uyxW2P.cu -o /tmp/uyxW2P.cu.sm_61.o`
command was executted for 2-3 seconds and no messages.
in /tmp i see `uyxW2P.cu` it contained `include <cuda_curand.h>`
in /tmp file was created `uyxW2P.cu.sm_61.o`  it' size 341960

**Comment by SlowBroshka** [comment]
No, I use clear ubntu on me desktop, no i don't have any ideas
Thanks your for your time

**Comment by maxicusj** [comment]
same issue here... CUDA 9.2 toolkit. manual compilation with the nvcc works. did you manage to resolve this?

**Comment by kaigai** [comment]
I setup Ubuntu18.04 + CUDA10 environment, however, could not reproduce the problem.

**Comment by maizarrahman** [comment]
This happened because pg_ctlcluster in (Ubuntu 18.04 or 16.04) does not read environment varibel in user profiles and pg_ctlcluster's environment file is empty.
As stated in /etc/init.d/postgresql, you must set PATH environment in /etc/postgresql/<version>/<cluster>/environment. 
For example, if you're using postgresql 9.6 with default installation, then it will be /etc/postgresql/9.6/main/environment. Add this to the file:
`PATH = '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'`

**Comment by SlowBroshka** [comment]
Thanks, that solved my problem (Ubuntu 18.04)
